tmt/base.py

```diff
@@ -1665,13 +1672,44 @@ def import_plan(self) -> Optional['Plan']:
                 else:
                     tmt.utils.git_clone(plan_id.url, destination, self)
                 if plan_id.ref:
-                    self.run(Command('git', 'checkout', plan_id.ref), cwd=destination)
+                    # attempt to evaluate dynamic reference
```

Comment:
if we are touching this function, could we remove one if on line 1657, rather return early ... we keep better line of sight? this is getting quite hard readable ...

Reply:
What do others think? @psss @happz I think that one recommendation is that final return value should be for a "default" case while `if`s in a function body should handle special cases. I really don't have an opinion myself but will obey the will of a majority.

Reply:
Hm, could be better, I'm not sure. I for one don't mind either variant.

Reply:
yeah :) Mostly what I would like to see is in case we are touching a lot functions like this, refactor them for better readability ideally, when we were reviewing this, it was hard to review, as there were too many if statements and it was hard to understand the flow of the function. 

Untested, but something like this would help, maybe also moving some parts of this long function to helper functions:

```
   def import_plan(self) -> Optional['Plan']:
          """ Import plan from a remote repository, return a Plan instance """
          if not self.is_remote_plan_reference:
              return None
  
          if self._imported_plan:
              return self._imported_plan
  
          assert self._remote_plan_fmf_id is not None  # narrow type
          plan_id = self._remote_plan_fmf_id
          self.debug(f"Import remote plan '{plan_id.name}' from '{plan_id.url}'.", level=3)
  
          # Use fmf cache for exploring plans (the whole git repo is not needed)
          if not self.my_run or self.my_run.opt('dry'):
  
              # fmf does not support dynamic ref so when working only with
              # the fmf cache we won't be able to read and process the dynamic_ref file
              # pointing us to the right branch.
              # The whole point of working with cache is to speed up things so it
              # probably doesn't make sense to clone whole repo.
              # Therefore, we have decided to lie here use a default branch instead.
              # An alternative would be either to enforce full clonning or to fail.
              if plan_id.ref and plan_id.ref.startswith('@'):
                  self.warn(
                      "Due to the 'dry' run tmt does not have data to evaluate dynamic "
                      f"reference used in the remote plan '{self.name} "
                      "Using the default branch instead.")
                  plan_id.ref = None
              try:
                  node = fmf.Tree.node(plan_id.to_minimal_spec())
              except fmf.utils.FetchError as error:
                  raise tmt.utils.GitUrlError(
                      f"Failed to import remote plan '{self.name}', "
                      f"use '--shallow' to skip cloning repositories.\n{error}")
  
              # Override the plan name with the local one to ensure unique names
              node.name = self.name
  
              # Create the plan object, save links between both plans
              self._imported_plan = Plan(node=node, run=self.my_run, logger=self._logger.descend())
              self._imported_plan._original_plan = self
  
              return self._imported_plan
  
          # Clone the whole git repository if executing tests (run is attached)
          assert self.parent is not None  # narrow type
          assert self.parent.workdir is not None  # narrow type
          destination = self.parent.workdir / "import" / self.name.lstrip("/")
  
          if plan_id.url is None:
              raise tmt.utils.SpecificationError(
                  f"No url provided for remote plan '{self.name}'.")
  
          if destination.exists():
              self.debug(f"Seems that '{destination}' has been already cloned.", level=3)
          else:
              tmt.utils.git_clone(plan_id.url, destination, self)
  
          if plan_id.ref:
              # attempt to evaluate dynamic reference
              try:
                  # Would it hurt if we simply rewrite plan_id.ref? Better choose safer path.
                  plan_id._ref_expanded = resolve_dynamic_ref(
                      workdir=destination,
                      ref=plan_id.ref,
                      plan=self,
                      logger=self._logger
                      )
              except tmt.utils.FileError as error:
                  raise tmt.utils.DiscoverError(
                      f"Failed to resolve dynamic ref of '{plan_id.ref}'.") from error
  
              if plan_id._ref_expanded:
                  self.run(
                      Command(
                          'git',
                          'checkout',
                          plan_id._ref_expanded),
                      cwd=destination)
  
          if plan_id.path:
              destination = destination / plan_id.path.unrooted()
  
          node = fmf.Tree(str(destination)).find(plan_id.name)
  
          # Override the plan name with the local one to ensure unique names
          node.name = self.name
  
          # Create the plan object, save links between both plans
          self._imported_plan = Plan(node=node, run=self.my_run, logger=self._logger.descend())
          self._imported_plan._original_plan = self
  
          return self._imported_plan
```

Reply:
> yeah :) Mostly what I would like to see is in case we are touching a lot functions like this, refactor them for better readability ideally, when we were reviewing this, it was hard to review, as there were too many if statements and it was hard to understand the flow of the function.

@thrix I can change it, however I believe that the "readability" part is quite subjective and what you are proposing in fact goes against [guidelines](https://docs.python-guide.org/writing/style/#returning-values) suggesting that a function should have one exit point. The thing is that despite you have one `if` statement less, in fact the reader needs to remember the negation of the condition anyway. You are not reducing the amount of conditions in the code flow, you would be just hiding it. What you are proposing basically means that a reader cannot focus on a specific block of code assuming it is "complete" because in order to understand its initial condition one has to review all previous blocks and assess in what way conditions used in these blocs are impacting the code flow.

Reply:
Interesting point about the multiple exit points. I'd naturally also minimize the number nested `if` statements to make the structure of the function as flat as possible. Anyway, here it seems the two cases (fmf cache versus full git clone) deserve a separate methods. This could resolve both readability and single (or very closely located) exit point.

Reply:
Interesting indeed, I for one tend to approach the code from the other side. I was thinking about it a few years ago, and functions I write are often filter-like: there's a thread, a goal to achieve, and along the way, I "filter out" invalid inputs, uncommon scenarios, side quests, and so on, until eventually, I have the pure gold :) Many exit points, but it's all clear in my head, often fairly shallow, with no deep nested `if`s or loops. Sort of a single corridor with many exits, without parallel corridors :)

```python
def push(guest, src, dst):
    if guest.guest is None:
        raise Exception()

    if guest.role != 'server':
        return

    if not src.exists():
        raise Exception()

   do_push(...)
```

```python
if self._imported_plan:
    return self._imported_plan

# build self._imported plan
...
return self._imported plan
```

^^ This is exactly what I tend to do - reduce surface one step at a time, and avoid nesting. I would never write `if not self._imported_plan` followed by an indented block of 68 lines. Other people would do this. I find it ugly and hard to follow. Other people find it nice and easier to follow. I don't know who's right.

I'd also try to flatten the design, and "crumble" it into smaller distinct steps, by introducing helper functions to be called from the main body. Easier to (unit) test, pure-ish, main function handles logic, helper functions take care of single one aspect of the whole picture. E.g. the following would be much more suited to how I ingest the code, despite me for one would not write it this way - but it's much easier for me to follow the logic flow, both helper methods can be unit tested, etc.

```python
        if not self._imported_plan:
            assert self._remote_plan_fmf_id is not None  # narrow type
            plan_id = self._remote_plan_fmf_id
            self.debug(f"Import remote plan '{plan_id.name}' from '{plan_id.url}'.", level=3)

            # Clone the whole git repository if executing tests (run is attached)
            if self.my_run and not self.my_run.opt('dry'):
                self._imported_plan = self._import_plan_from_clone()

            # Use fmf cache for exploring plans (the whole git repo is not needed)
            else:
                self._imported_plan = self._import_plan_from_fmf_cache()

           # Override the plan name with the local one to ensure unique names
            node.name = self.name
            # Create the plan object, save links between both plans
            self._imported_plan = Plan(node=node, run=self.my_run, logger=self._logger.descend())
            self._imported_plan._original_plan = self

        return self._imported_plan
```

Reply:
Completely agree with @happz here, splitting into smaller functions would definitely help with readability but let's not block the pull request on this for now.